pkg/watcher/watcher.go

```diff
@@ -0,0 +1,62 @@
+package watcher
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+// GetWatchCMMount returns a function that watches the provided path mounted from a
+// configmap and runs the eventFunc function whenever the configmap is updated
+// until the provided context is cancelled. If an error occurs, errFunc is run.
+// An error is returned if a watcher could not be created or set to watch the
+// provided path or if the parovided path is not a configmap mount.
+func GetWatchCMMount(path string, eventFunc func() error, errFunc func(error)) (func(ctx context.Context), error) {
+	// Verify that this is the root of a configmap mount
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, fmt.Errorf("Could not read provided directory %s: %v", path, err)
+	}
+	isCMMount := false
+	for _, file := range files {
+		if file.Name() == "..data" {
+			isCMMount = true
+		}
+	}
+	if !isCMMount {
+		return nil, fmt.Errorf("Provided directory %s is not a configmap directory", path)
+	}
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	err = w.Add(path)
+	if err != nil {
+		return nil, err
+	}
+	return func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				if err := w.Close(); err != nil {
+					errFunc(err)
+					//logrus.WithError(err).Error("Failed to close fsnotify watcher")
+				}
+				return
+			case event := <-w.Events:
+				if event.Name == "..data" && event.Op == fsnotify.Create {
```

Comment:
The steps that occur during a configmap update are:
1. A new timestamped directory is created with a `..` prefix
2. New symlinks (that point to `..data/<filename>`) are created if necessary
3. A `..data_tmp` symlink that points to the timestamped directory is made
4. `..data_tmp` is moved to `..data`, in a way that results in a `CREATE` event for `..data`.
5. Remove stale (deleted) symlinks

Since the last important step is the `CREATE` event on `..data`, we can run reloads when `..data` gets a `CREATE` event. The newly added integration tests simulate part of this behavior.